design-proposals/klip-30-lambda-functions.md

```diff
@@ -0,0 +1,121 @@
+
+## Public APIs
+
+There are two aspects of the user-facing lambda function interface: syntax for describing the actual lambda functions, and a way to specify how to apply these functions to input values.
+
+## Design
+
+### Syntax
+
+It would probably be ideal to use Java-style lambda syntax, but we already use the `->` symbol to represent `Struct` field lookups. Giving `->` a double meaning (if that’s even possible) would create ambiguity and complexity in the grammar/parser. We therefore propose using the `=>` symbol to signify a lambda function:
+
+```
+arg => expr
+```
+
+This expression represents the following:
+
+* `arg` is the argument passed into the scope of each invocation of the lambda function.
+* `expr` is the "body" of the lambda function. Each invocation of the lambda function will evaluate this expression and return the result.
+
+Multiple arguments should also be allowed for lambda functions. When multiple arguments are used, they should be wrapped in parentheses:
+
+```
+(x, y) => expr
+```
+
+#### Allowed expressions
+
+The body of each lambda function should simply be **any expression that could be used in a `SELECT` expression list**, with some exceptions:
+
+* Aggregates should not be allowed (please see *[Open questions](#open-questions)* section)
+* `SELECT` subqueries should not be allowed
+
+#### Examples
+
+Given these restrictions, the following examples would be valid lambda functions:
+
+* `x => x + 1`
+* `s => UCASE(s)`
+* `(x, y) => x + y`
+* `(x, y) => CASE WHEN x IS NULL THEN y ELSE x + y END`
+* `(x, y) => x + COALESCE(y, 0)`
+
+### Invocation functions
+
+Lambda functions should require a specific **invocation function**, which tells ksqlDB how to apply the given lambda function to an input value, and what kind of result to return. The following invocation functions are proposed for the initial support for lambda functions:
+
+* `transform_array(arr, x => y)` - Applies the given lambda function to each element of the input `Array`, returning a new `Array` containing the transformed output.
+
+* `transform_map(map, (k, v) => x, y)` - Applies the given lambda function to each key-value pair of the input `Map`, returning a new `Map` containing the transformed key-value pairs.
+
+* `filter_array(arr, x => bool)` - Filters the input `Array` using the given lambda function. A new `Array` is returned, containing only values for which the lambda function evaluated to `true`.
+
+* `filter_map(map, (k, v) => bool)` - Filters the input `Map` using the given lambda function. A new `Map` is returned, containing only the key-value pairs for which the lambda function evaluated to `true`.
+
+* `reduce_array(arr, s, (x, s) => s)` - Reduces the input `Array` down to a single value. `s` is the initial state and is passed into the scope of the lambda function. Each invocation returns a new value for `s`, which the next invocation will receive. `reduce_array` will return the final value of `s`.
```

Comment:
Above you state: `Aggregates should not be allowed` -- seems to be a contradiction?

Reply:
By aggregates I meant aggregate functions such as `sum`, `count`, `avg` etc. I will clarify since you are correct that `reduce_*` is a kind of aggregate.

Reply:
I clarified the restriction in question.

Reply:
Could you elaborate a bit why `reduce` would be allowed but other `sum/count/avg` etc would not? I think implementation wise they would be the same. And if we are concerned about inferring the return data type the other `transform_map` would already have this problem anyways.